tests/integration/test_core.py

```diff
@@ -0,0 +1,124 @@
+
+
+@pytest.mark.asyncio
+async def test_openstack_object_storage_plugin(  # noqa: C901
+    ops_test: pytest_operator.plugin.OpsTest,
+    unit_ip_list,
+    openstack_environment,
+):
+    """
+    arrange: after charm deployed, db relation established and openstack swift server ready.
+    act: update charm configuration for openstack object storage plugin.
+    assert: openstack object storage plugin should be installed after the config update and
+        WordPress openstack swift object storage integration should be set up properly.
+        After openstack swift plugin activated, an image file uploaded to one unit through
+        WordPress media uploader should be accessible from all units.
+    """
+    swift_conn = swiftclient.Connection(
+        authurl=openstack_environment["OS_AUTH_URL"],
+        auth_version="3",
+        user=openstack_environment["OS_USERNAME"],
+        key=openstack_environment["OS_PASSWORD"],
+        os_options={
+            "user_domain_name": openstack_environment["OS_USER_DOMAIN_ID"],
+            "project_domain_name": openstack_environment["OS_PROJECT_DOMAIN_ID"],
+            "project_name": openstack_environment["OS_PROJECT_NAME"],
+        },
+    )
+    container_exists = True
+    container = "content-cache"
+    try:
+        swift_conn.head_container(container)
+    except swiftclient.exceptions.ClientException as e:
+        if e.http_status == 404:
+            container_exists = False
+        else:
+            raise e
+    if container_exists:
+        for swift_object in swift_conn.get_container(container, full_listing=True)[1]:
+            swift_conn.delete_object(container, swift_object["name"])
+        swift_conn.delete_container(container)
+    swift_conn.put_container(container)
+    app = ops_test.model.applications["content-cache-k8s"]
+    await app.set_config({"backend": f"http://{swift_conn.url}:80"})
+    await app.set_config({"site": swift_conn.url})
+    swift_service = swiftclient.service.SwiftService(
+        options=dict(
+            auth_version="3",
+            os_auth_url=openstack_environment["OS_AUTH_URL"],
+            os_username=openstack_environment["OS_USERNAME"],
+            os_password=openstack_environment["OS_PASSWORD"],
+            os_project_name=openstack_environment["OS_PROJECT_NAME"],
+            os_project_domain_name=openstack_environment["OS_PROJECT_DOMAIN_ID"],
+        )
+    )
+    swift_service.post(container=container, options={"read_acl": ".r:*,.rlistings"})
+
+    for idx, unit_ip in enumerate(unit_ip_list):
+        nonce = secrets.token_hex(8)
+        filename = f"{nonce}.{unit_ip}.{idx}"
+        for r in swift_service.upload(
+            container=container,
+            objects=[swiftclient.service.SwiftUploadObject(None, object_name=filename)],
+        ):
+            if r["success"]:
+                if "object" in r:
+                    print(r["object"])
+                elif "for_object" in r:
```

Comment:
Do you think it would be worth reducing the amount of nested code here?

Reply:
that is a debugging code I forgot to delete, so it will be gone :)